lib/nrf_modem_lib/nrf91_connectivity/nrf91_connectivity.c

```diff
@@ -0,0 +1,299 @@
+void pdn_event_handler(uint8_t cid, enum pdn_event event, int reason)
+{
+	ARG_UNUSED(cid);
+
+	int ret;
+
+	switch (event) {
+	case PDN_EVENT_CNEC_ESM:
+		LOG_DBG("Event: PDP context %d, %s", cid, pdn_esm_strerror(reason));
+		break;
+	case PDN_EVENT_ACTIVATED:
+		LOG_DBG("PDN connection activated");
+		LOG_DBG("PDN IPv4 up");
+
+		ret = ipv4_addr_add(iface_bound);
+		if (ret) {
+			LOG_ERR("ipv4_addr_add, error: %d", ret);
+			net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+			nrf91_connectivity_disconnect(NULL);
+			return;
+		}
+
+		net_if_dormant_off(iface_bound);
+
+		/* Cancel ongoing LTE timeout work, if ongoing. */
+		k_work_cancel_delayable(&lte_timeout_work);
+
+		break;
+	case PDN_EVENT_DEACTIVATED:
+		LOG_DBG("PDN connection deactivated");
+		LOG_DBG("PDN IPv4 down");
+
+		ret = ipv4_addr_remove(iface_bound);
+		if (ret) {
+			LOG_ERR("ipv4_addr_remove, error: %d", ret);
+			net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+			net_if_dormant_on(iface_bound);
+			return;
+		}
+
+		ret = ipv6_addr_remove(iface_bound);
+		if (ret) {
+			LOG_ERR("ipv6_addr_remove, error: %d", ret);
+			net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+			net_if_dormant_on(iface_bound);
+			return;
+		}
+
+		net_if_dormant_on(iface_bound);
+
+		/* If persistence is disabled, LTE is deactivated upon a lost connection.
+		 * Re-establishment is reliant on the application calling net_if_connect()
+		 * or net_if_up() if NET_IF_NO_AUTO_CONNECT is disabled to
+		 * re-establish the LTE connection.
+		 */
+		if (!net_if_get_conn_persistence(iface_bound)) {
+			ret = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_DEACTIVATE_LTE);
+			if (ret) {
+				LOG_ERR("lte_lc_func_mode_set, error: %d", ret);
+				net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR,
+						      iface_bound);
+				nrf91_connectivity_disconnect(NULL);
+				return;
+			}
+		}
+
+		break;
+	case PDN_EVENT_IPV6_UP:
+		LOG_DBG("PDN IPv6 up");
+
+		ret = ipv6_addr_add(iface_bound);
+		if (ret) {
+			LOG_ERR("ipv6_addr_add, error: %d", ret);
+			net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+			nrf91_connectivity_disconnect(NULL);
+			return;
+		}
+
+		break;
+	case PDN_EVENT_IPV6_DOWN:
+		LOG_DBG("PDN IPv6 down");
+
+		ret = ipv6_addr_remove(iface_bound);
+		if (ret) {
+			LOG_ERR("ipv6_addr_remove, error: %d", ret);
+			net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+			net_if_dormant_on(iface_bound);
+			return;
+		}
+
+		break;
+	default:
+		LOG_ERR("Unexpected PDN event: %d", event);
+		break;
+	}
+}
+
+/* Public APIs */
+void nrf91_connectivity_init(const struct net_if_conn *if_conn)
+{
+	struct net_if_conn *if_conn_cast = (struct net_if_conn *)if_conn;
+
+	/* Set default values for the network interface's auto options. */
+	if (IS_ENABLED(CONFIG_NRF91_CONNECTIVITY_NET_IF_NO_AUTO_START)) {
+		net_if_flag_set(if_conn->iface, NET_IF_NO_AUTO_START);
```

Comment:
Haven't checked where the naming comes from, perhaps upstream there were already similarly named settings, but `_NO_AUTO_START` is confusing.

It is not at all immediate to understand what these checks do:
```
IS_ENABLED(_NO_AUTO_START) -> doesn't start automatically
!IS_ENABLED(_NO_AUTO_START) -> starts automatically
```
these would have been much easier to understand:
```
IS_ENABLED(_AUTO_START) -> starts automatically
!IS_ENABLED(_AUTO_START) -> does not start automatically
```

Reply:
same for `_CONNECT`.
Perhaps since we have our own Kconfig entries downstream, it might be smart to reword those so they are simpler for the users of this library to understand.

Reply:
Yes 😢 I completely agree. I followed the naming convention used upstream to keep naming close to the source.
I'm more than happy to change this if there are no objections. cc @rlubos

Reply:
Downstream you can name it however you like, as long as it's reasonable. But for upstream, I don't think this would be feasible to change, given that the flag is used widely, could also be used out-of-tree, IMO not worth it. Especially, that I personally don't find it **that** confusing.

Reply:
Keeping it as is for now to align with upstream.